### ISSUE TITLE
False positive when using super()

### ISSUE THREAD

**Issue by ivanlonel** [atollk/flake8-classmethod-staticmethod#3]
When you override a class method, you may call `super()` instead of referencing `cls`, which wouldn't be allowed in a static method.

So, maybe class methods like the one below shouldn't trigger `CSM131`?

```python
class LoginSpider(scrapy.Spider):

    @classmethod
    def from_crawler(cls, crawler, *args, **kwargs):
        spider = super().from_crawler(crawler, *args, **kwargs)
        crawler.signals.connect(spider.spider_opened, scrapy.signals.spider_opened)
        return spider

    ...
```

**Comment by atollk** [comment]
Interesting, I didn't know that's a thing. 

So would the fix be to look for any usages of the `super` function in addition to usages of the `cls` parameter? Or would that miss anything else?

**Comment by ivanlonel** [comment]
As far as I understand it, and disregarding possible shady language extrapolations (like using custom method-altering decorators) this should be the only exception to requiring use of `cls` in the method body.

In fact, you are referencing `cls`, just implicitly. The line where we call `super()` in my example above is semantically identical to `spider = super(LoginSpider, cls).from_crawler(crawler, *args, **kwargs)`. And writing it in this form wouldn't trigger `CSM131`. But that defeats perhaps the main advantage of a class method over a static method: not needing to reference the class by name (and the parent method is already a class method, so we couldn't override it as static anyway).

And no, `super(cls, cls)` wouldn't be correct, as explained in [this great SO answer on the subject](https://stackoverflow.com/a/1817840/3964167).

**Comment by ivanlonel** [comment]
By the way, calling `super` with arguments should still trigger CSM131 if `cls` is not referenced, since that could also be done in a static method. Only zero-argument super() magically infers the class it's being called in.

So I believe the check should look specifically for `cls` and `super()` without arguments.

**Comment by atollk** [comment]
> So I believe the check should look specifically for cls and super() without arguments.

But `super(A, cls)` would also be a valid use, wouldn't it? If my understanding is correct, any call to `super` with an argument of `cls` should be exempt from the rule.

I understand that `super(cls, cls)` is erroneous code, but either someone is doing some black Python magic and it is intended, or it is an obvious bug that should be find by a basic test. Either way, I feel like it should not be reported either by the rule, as that call is not something that could be expressed in a `staticmethod`.

**Comment by ivanlonel** [comment]
I agree with both affirmations, but the thing is `super(A, cls)` and even `super(cls, cls)` are already exempt because they include references to `cls`, which is enough to make `CSM131` happy as it is today. I mean, checking "`cls` and `super()` without arguments" includes these cases, not because of the `super` part, but because of `cls` being explicitly passed as argument.

That's why on my previous comment I said:

> calling `super` with arguments should still trigger CSM131 if `cls` is not referenced

emphasis on the "if". :)

In other words, calling `super` should not be enough to avoid `CSM131` when it's called with arguments and `cls` is not among these arguments. But since your current implementation already takes care of `cls`, I believe adding a check for `super()` without arguments should be enough to only raise `CSM131` when the method could actually be made static without loss.

**Comment by atollk** [comment]
Alright, sounds good. I'll see that I can add the functionality within this week.

**Comment by atollk** [comment]
@ivanlonel Maybe you can check out #4 and if that looks good to you, I'll merge it and publish a new version.

**Comment by ivanlonel** [comment]
Looks great to me! This takes care of the false positive.
Thank you so much for your attention, time and effort, @atollk.
